File: src/strings/InsertInterval_57.java
package strings;

import java.util.ArrayList;
import java.util.List;

import static array.MaximalSquare_221.printDim;

public class InsertInterval_57 {

    public static void main(String[] args) {
        System.out.println(printDim(new InsertInterval_57().insert(new int[][]{{1, 3}, {6, 9}}, new int[]{2, 5})));
        System.out.println(printDim(new InsertInterval_57().insert(new int[][]{{1, 2}, {3, 5}, {6, 7}, {8, 10}, {12, 16}}, new int[]{4, 8})));
    }


    public int[][] insert(int[][] intervals, int[] newInterval) {
        int len = intervals.length;
        List<int[]> res = new ArrayList<>();

        int newStart = newInterval[0];
        int newEnd = newInterval[1];

        boolean merged = false;

        for (int i = 0; i < len; i++) {
            int[] interval = intervals[i];
            int start = interval[0];
            int end = interval[1];
            if (end < newStart) {
                res.add(interval);
            } else if (start > newEnd) {
                if (!merged) {
                    res.add(new int[]{newStart, newEnd});
                    merged = true;
                }
                res.add(interval);
            } else {
                newStart = Math.min(newStart, start);
                newEnd = Math.max(newEnd, end);
            }
        }

        if (!merged) {
            res.add(new int[]{newStart, newEnd});
        }

        return res.toArray(new int[0][0]);
    }
}

File: src/merge_intervals/insert_interval_57.py
# 1. sorted - done
# 2. insert by start time O(N)
# 3. merge O(N)
from typing import List


class Solution:
    def insert(self, intervals: List[List[int]], new_interval: List[int]) -> List[List[int]]:
        if len(intervals) == 0:
            return [new_interval]
        merged = []

        i, start, end = 0, 0, 1

        while i < len(intervals) and intervals[i][end] < new_interval[start]:
            merged.append(intervals[i])
            i += 1

        while i < len(intervals) and intervals[i][start] <= new_interval[end]:
            new_interval[start] = min(intervals[i][start], new_interval[start])
            new_interval[end] = max(intervals[i][end], new_interval[end])
            i += 1

        merged.append(new_interval)

        while i < len(intervals):
            merged.append(intervals[i])
            i += 1

        return merged


def main():
    print("Intervals after inserting the new interval: " + str(Solution().insert([[1, 5]], [2, 7])))
    print("Intervals after inserting the new interval: " + str(Solution().insert([[1, 3], [5, 7], [8, 12]], [4, 6])))
    print("Intervals after inserting the new interval: " + str(Solution().insert([[1, 3], [5, 7], [8, 12]], [4, 10])))
    print("Intervals after inserting the new interval: " + str(Solution().insert([[2, 3], [5, 7]], [1, 4])))
    print("Intervals after inserting the new interval: " + str(Solution().insert([], [1, 4])))


main()
